### ISSUE TITLE
Poor documentation leads to obscure bug where JSON authentication leads to 404 errors in modern runtimes such as FrankenPHP

### ISSUE THREAD

**Issue by dwgebler** [api-platform/core#6831]
This issue is minimally reproducible by using the latest API Platform distribution template. You will need to run the production Docker config to use FrankenPHP in worker mode.

Add the Lexik JWT authentication bundle to a new project and configure an auth endpoint as per normal.

Add an API resource with some validation constraint on it.

Use the following configuration for `packages/api_platform.yaml`:

```yaml
api_platform:
    formats:
        json: ['application/json']
        jsonld: ['application/ld+json']
        multipart: ['multipart/form-data']
    defaults:
        formats:
            json: ['application/json']
            jsonld: ['application/ld+json']
            multipart: ['multipart/form-data']
        extra_properties:
            rfc_7807_compliant_errors: true
```

Make a request to your auth endpoint and obtain a token, authenticate with this token against a request to your API resource to a POST endpoint, but in the request body, violate your validation constraint to return a 422 response.

Now immediately call your auth endpoint - you will get a 404 error, because Symfony will determine the authenticator doesn't support the request - and because the auth endpoint doesn't have a real controller, when the authenticator doesn't run, no controller will be resolved and the route won't be matched to anything. This is happening because in the long-running worker, the 422 response has returned `application/problem+json` and overwritten the request format mimetype mappings (which are static) in the HttpFoundation `Request` class probably by calling `setFormat` somewhere.

The fix is to use the corrected config like below:

```yaml
api_platform:
    formats:
        json: ['application/json', 'application/problem+json']
        jsonproblem: [ 'application/json', 'application/problem+json' ]
        jsonld: ['application/ld+json']
        multipart: ['multipart/form-data']
    defaults:
        formats:
            json: ['application/json', 'application/problem+json']
            jsonproblem: [ 'application/json', 'application/problem+json' ]
            jsonld: ['application/ld+json']
            multipart: ['multipart/form-data']
        extra_properties:
            rfc_7807_compliant_errors: true
```

But the documentation for API Platform on content negotiation and error handling really isn't clear. This caused a bug for me in a production environment that took ages to track down and gave me a headache, so for the benefit of others in future, please address this somewhere. Thanks.

**Comment by dwgebler** [comment]
Yeah see https://github.com/api-platform/core/blob/de3bed269838556d46899e78ed5603b7e6229e8b/src/State/Provider/ContentNegotiationProvider.php#L54 this isn't _adding_ the configured MIME types, it's _replacing_ them. I don't think this one is on Symfony to change, because the way it works is the apparent intended behaviour. 

@dunglas I think API Platform needs to ensure that the default MIME types for any known types are always included here, something like:

```php
    private function addRequestFormats(Request $request, array $formats): void
    {
        foreach ($formats as $format => $mimeTypes) {
            if ($format == 'json' && !in_array('application/json', $mimeTypes, true)) {
                $mimeTypes[] = 'application/json';
            }
            if ($format === 'jsonld' && !in_array('application/ld+json', $mimeTypes, true)) {
                $mimeTypes[] = 'application/ld+json';
            }
            $request->setFormat($format, (array) $mimeTypes);
        }
    }
```

So I would say it's an API Platform bug to fix.

**Comment by dwgebler** [comment]
@soyuka yeah the issue is that without the explicit formats configuration, in the `ContentNegotiatorProvider`, you have these lines in `provide(...)`:

```php
   public function provide(Operation $operation, array $uriVariables = [], array $context = []): object|array|null
    {
        if (!($request = $context['request'] ?? null) || !$operation instanceof HttpOperation) {
            return $this->decorated?->provide($operation, $uriVariables, $context);
        }

        $isErrorOperation = $operation instanceof ErrorOperation;

        $formats = $operation->getOutputFormats() ?? ($isErrorOperation ? $this->errorFormats : $this->formats);
        $this->addRequestFormats($request, $formats);
...
```

The operation output formats from `getOutputFormats()` will default to `['json' => ['application/problem+json']]` which is then passed to `addRequestFormats`.

So as it is now, that code will result in the `Request::$formats` array replacing the supported MIMEs for format `json` with just `['application/problem+json']` meaning a subsequent request on a long-running worker will be unable to match a request for `application/json` to the `json` format. Explicitly configuring the formats in API Platform resolves the issue, but either the need to do this should be documented or - as I would say this behaviour is a bug - the correct IANA MIME types should always be added to the arrays passed to `Request::setFormat`.

The call to `$request->setFormat($format, (array) $mimeTypes)` with parameters `'json', ['application/problem+json']` is the cause of the bug I experienced.

**Comment by soyuka** [comment]
`outputFormats` contains `application/problem+json` only when an Error occurs. I don't understand how subsequent request would keep formats from a previous request as this is quite dangerous, and its why we always reset the request formats. 

IIUC you have:

1. an error occurs => `addRequestFormats($request, 'application/problem+json)'
2. a subsequent request occurs and throws a 406 because the `problem+json` format is not supported

Did I got this right?

**Comment by dwgebler** [comment]
@soyuka not quite.

A validation error triggers a call to addRequestFormats with `['json' => ['application/problem+json']]`

A subsequent `application/json` request to the check path of a JsonLoginAuthenticator such as Lexik JWT, where API Platform is running via a worker model such as FrankenPHP, will fail with a 404 because Symfony will not recognise it can support the request MIME type, because there is no longer a mapping of `json` to `application/json` in `Request` class.

For clarity, `Request::setFormat()` does _not_ set the format of an individual request instance, it sets the mapping of formats to MIME types that are supported by all Requests.

**Comment by soyuka** [comment]
`Request::setFormat` should be called on every request and we should not rely on previous values.

**Comment by dwgebler** [comment]
> `Request::setFormat` should be called on every request and we should not rely on previous values.

That's not going to work, because there's no way [in `Request`] to reset the values of $formats back to their initial state, because it's a static property. So API Platform should either ensure that recognised IANA types for the formats it supports are always included in the calls to `setFormat` in the ContentNegotiationProvider or it needs to make much clearer in the docs that it's necessary to configure _all_ the mime types for each format yourself if you're not using the default configuration for `api_platform.formats` - nothing in the docs suggests to a user that they would need to put both application/json and application/problem+json in the list of MIME types for json and jsonproblem if they're changing those defaults.

**Comment by dwgebler** [comment]
@dunglas @soyuka I have made a minimal reproducer here https://github.com/dwgebler/ap-error-test

**Comment by soyuka** [comment]
> That's not going to work, because there's no way [in Request] to reset the values of $formats back to their initial state, because it's a static property.

I think that this is an issue. Setting all the formats is an issue as there may be different formats configured across different operations on the same API.

**Comment by dwgebler** [comment]
@soyuka @dunglas see proposed fix #6833

**Comment by soyuka** [comment]
We still want this to be fixed inside Symfony I've opened an issue there.
